src/components/WithdrawalCredentials.tsx

```diff
@@ -0,0 +1,321 @@
+  @media (max-width: ${screenSizes.small}) {
+    width: 100%;
+  }
+`;
+
+const ResultsLineItem = styled.div`
+  display: flex;
+  gap: 0.5rem;
+  @media (max-width: ${screenSizes.small}) {
+    flex-direction: column;
+  }
+  &:not(:last-of-type) {
+    margin-bottom: 1rem;
+  }
+`;
+
+const ResultsLineItemDesktopOnly = styled(ResultsLineItem)`
+  @media (max-width: ${screenSizes.medium}) {
+    display: none;
+  }
+`;
+
+const ResultLabel = styled.div`
+  white-space: nowrap;
+  font-weight: bold;
+`;
+
+const AddressCopyContainer = styled.div`
+  gap: 1rem;
+  display: flex;
+  flex-wrap: nowrap;
+  overflow: hidden;
+`;
+
+const LongAddress = styled.div`
+  white-space: nowrap;
+  overflow: hidden;
+  text-overflow: ellipsis;
+`;
+
+const CopyContainer = styled.button`
+  font-family: monospace;
+  width: fit-content;
+  font-size: 1rem;
+  padding: 0 5px;
+  cursor: pointer;
+  white-space: nowrap;
+  background: rgb(255 255 255 / 50%);
+  letter-spacing: 1.5px;
+  border: 1px solid ${p => p.theme.blue.dark};
+  border-radius: 4px;
+  &:hover {
+    box-shadow: 0px 8px 17px rgba(0, 0, 0, 0.15);
+    background-image: linear-gradient(to right, ${p => p.theme.rainbow});
+  }
+  &:focus {
+    outline-offset: -2px;
+  }
+`;
+
+interface Validator {
+  validatorIndex: number;
+  withdrawalCredentials: string;
+  isUpgraded: boolean;
+}
+
+interface IProps {}
+export const WithdrawalCredentials: FC<IProps> = () => {
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  const [hasError, setHasError] = useState<boolean>(false);
+  const [inputValue, setInputValue] = useState<string>('');
+  const [validator, setValidator] = useState<Validator | null>(null);
+  const [copied, setCopied] = useState<{ [key: string]: boolean }>({});
+  const isMobile: boolean = (window as any).mobileCheck();
+
+  const checkWithdrawalCredentials = async () => {
+    setHasError(false);
+    setIsLoading(true);
+    const endpoint = `${BEACONCHAIN_URL}/api/v1/validator/${inputValue}`;
+    try {
+      const response = await fetch(endpoint);
+      const { data } = await response.json();
+      const withdrawalCredentials = data.withdrawalcredentials;
+      setValidator({
+        validatorIndex: parseInt(inputValue, 10),
+        withdrawalCredentials,
+        isUpgraded: withdrawalCredentials.startsWith('0x01'),
+      });
+    } catch (error) {
+      setHasError(true);
+    } finally {
+      setIsLoading(false);
+    }
+  };
+  const onCopy = (key: string) => {
+    setCopied(prev => ({ ...prev, [key]: true }));
+    setTimeout(() => {
+      setCopied(prev => ({ ...prev, [key]: false }));
+    }, 2000);
+  };
+
+  const handleChange = (e: ChangeEvent<HTMLInputElement>) =>
+    setInputValue(e.target.value.replace(/\D/g, ''));
+
+  const longAddress = useMemo<string>(
+    () => (validator ? `0x${validator.withdrawalCredentials.slice(-40)}` : ''),
+    [validator]
+  );
+
+  const resultText = useMemo<string | JSX.Element>(() => {
```

Comment:
Would not recommend using `useMemo` for this type of logic. The purpose of `useMemo` is to memoize an expensive calculation.

The logic inside looks like a common react flow. In case you want to abstract this from this component, I'd suggest creating a new component instead.

```tsx
function ResultText({ isLoading, hasError }) {
  if (isLoading) return <Spinner />;
  if (hasError) ...

  return ...
}
```

====
If there is a nested expensive calculation in this tree, I'd suggest using `useMemo` only for that specific thing.